engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java

```diff
@@ -0,0 +1,135 @@
+package org.enso.interpreter.node.expression.builtin.meta;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.Truffle;
+import com.oracle.truffle.api.TruffleFile;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import java.util.Optional;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.node.EnsoRootNode;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+import org.enso.interpreter.runtime.data.EnsoFile;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.error.DataflowError;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+import org.enso.pkg.Package;
+
+@BuiltinMethod(
+    type = "Project_Description",
+    name = "enso_project_builtin",
+    description = "Returns the project description of the project given as the argument")
+public abstract class EnsoProjectNode extends Node {
+  public static EnsoProjectNode build() {
+    return EnsoProjectNodeGen.create();
+  }
+
+  /**
+   * @param module Either {@code Nothing}, or a module.
+   */
+  public abstract Object execute(Object module);
+
+  /**
+   * Fetches the second stack frame from Truffle runtime (note that the first stack frame is the
+   * call of {@code enso_project}) - the caller of {@code enso_project}, and finds in which package
+   * the caller is located.
+   *
+   * @param nothing Nothing, or interop null.
+   */
+  @Specialization(guards = "isNothing(interop, nothing)")
+  @TruffleBoundary
+  public Object getCurrentProjectDescr(
+      Object nothing, @CachedLibrary(limit = "5") InteropLibrary interop) {
+    var ctx = EnsoContext.get(this);
```

Comment:
TCR: The result of this method is always the same (as the node is inserted at single source location)  for every `ctx` - cache it and avoid `@TruffleBoundary` on subsequent invocations.

Reply:
Add two `@CompilationFinal` fields into the `EnsoProjectNode`:
```
@CompilationFinal
private Reference<EnsoContext> previousCtxt;
@CompilationFinal
private Reference<Atom> previousProject;
```
and speculate on `EnsoContext` being always the same. Then the result shall also be the same.

Reply:
Done in 9355e70d8da2721b6705289ef31e72e3cc88dcde